### PR TITLE
[#369] Inactive migration plans can be scheduled to run 

### LINKED ISSUE #369: User can schedule a migration plan
The user should be able to schedule a migration plan, by specifying a date and time for it to be run.

### PR DESCRIPTION
**Pull request by AllenBW**
fixes #369

### Where we're at... 
yes

this one's dedicated to @bdunne @gtanzillo @abellotti

### CODE REVIEW

**Comment by AllenBW** [comment]
@vconzola any chance yah have the rest of the mockups for this available?  [(as mentioned here)](https://github.com/ManageIQ/miq_v2v_ui_plugin/issues/369#issuecomment-395788834)


### Other things we're waiting on

- [ ] Theres a lil api work @bdunne is workin' on... 
- [ ] @serenamarie125 @vconzola any word on what we're doing for a date time picker here? or for now just use something, anything, even if separate to get this in 
- [ ] profit

🌮 💃

File: app/javascript/react/screens/App/Overview/components/Migrations/MigrationsNotStartedList.js [hunk 1]
```diff
@@ -1,78 +1,166 @@
...
+          <ListView className="plans-not-started-list" style={{ marginTop: 0 }}>
+            {notStartedPlans.map(plan => {
+              const confirmationWarningText = (
+                <React.Fragment>
+                  <p>
+                    {__('Are you sure you want to unschedule plan ')}
+                    <strong>{plan.name}</strong>
+                    {__(' targted to run on ')}
+                    {moment(plan.schedule_time).format('MMMM Do YYYY, h:mm a')}
+                    {__('?')}
```
Comment (AparnaKarve) [inline app/javascript/react/screens/App/Overview/components/Migrations/MigrationsNotStartedList.js]: I am aware the PR is still WIP, but here's the thing that we discussed in the i18n context about complete sentences.

The sentence should ideally look like this -
"Are you sure you want to unschedule plan %s targeted to run on %s?"

Comment (AparnaKarve) [reply app/javascript/react/screens/App/Overview/components/Migrations/MigrationsNotStartedList.js]: Check out https://github.com/ManageIQ/miq_v2v_ui_plugin/blob/9aa142a0b212750cb7affb286654e0c98b49c1e9/app/javascript/react/screens/App/Overview/components/DeleteInfrastructureMappingConfirmationModal/DeleteInfrastructureMappingConfirmationModal.js#L58 for styled, `sprintf` implementation.

Comment (AllenBW) [reply app/javascript/react/screens/App/Overview/components/Migrations/MigrationsNotStartedList.js]: yeah! thanks for the 👁 🙇 ❤️

File: app/javascript/react/screens/App/Overview/components/Migrations/MigrationsNotStartedList.js [hunk 2]
```diff
@@ -1,78 +1,168 @@
...
+            {notStartedPlans.map(plan => {
+              const confirmationWarningText = (
+                <React.Fragment>
+                  <p>
+                    {sprintf(
+                      __(
+                        'Are you sure you want to unschedule plan %s  targted to run on %s ?'
+                      ),
+                      plan.name,
+                      moment(plan.schedule_time).format('MMMM Do YYYY, h:mm a')
```
Comment (priley86) [inline app/javascript/react/screens/App/Overview/components/Migrations/MigrationsNotStartedList.js]: can this one use our `formatDateTime` helper? 🍬

File: app/javascript/react/screens/App/Overview/components/Migrations/MigrationsNotStartedList.js [hunk 3]
```diff
@@ -1,78 +1,168 @@
...
+                    </ListView.InfoItem>,
+                    !plan.schedule_time && (
+                      <ListView.InfoItem
+                        key={plan.id + 1}
+                        style={{ textAlign: 'left' }}
+                      >
+                        {__(`Migration scheduled`)}
+                        <br />
+                        {moment(plan.schedule_time).format(
+                          'MMMM Do YYYY, h:mm a'
```
Comment (priley86) [inline app/javascript/react/screens/App/Overview/components/Migrations/MigrationsNotStartedList.js]: same deal w/ the helper

**Comment by AllenBW** [comment]
Just in case an inquiring mind was wondering... ux is done.... our path to completion is presently
👇 this pr has to be merged:
https://github.com/ManageIQ/manageiq/pull/17594
so 👉  this pr can be merged: https://github.com/ManageIQ/manageiq-api/pull/400

then whatever cleanup befalls after the fact... 

profit 🌮 🥇

**Comment by AllenBW** [comment]
I have another update on this. (*TL;DR* we're blocked)

 Both prs needed for this work have been merged 🎊 🎉 🙌 

*BUT* these things... don't quite do everything we need, such as [(this is where the work lives btw)
](https://github.com/ManageIQ/manageiq/blob/master/app/models/service_template.rb#L409):

1. Ability to update (edit) a service template schedule, presently scheduling when one is already present results in 
```
{
  "error": {
    "kind": "internal_server_error",
    "message": "Validation failed: Name has already been taken",
    "klass": "ActiveRecord::RecordInvalid"
  }
}
``` 
2. View the scheduled model on `/api/service_template/#id`, `scheduled` is not available as a virtual attribute or a relationship
3. Ability to delete a schedule, but this might fall under edit with a null or something...

@gtanzillo, the generous wonderful fella that got this work started is now enjoying some well deserved vacay, was wondering there existed other resources that could tackle this?

**Comment by gtanzillo** [comment]
@AllenBW Schedule update and delete have not been completed as of yet, that's still a work in progress.

**Comment by AllenBW** [comment]
Theres relief in sight, https://github.com/ManageIQ/manageiq-api/pull/408

**Comment by AllenBW** [comment]
So as you might have noticed, [datetimepicker has made it into patternfly 
](https://github.com/patternfly/patternfly/releases/tag/v3.52.0)

but before we can use it here, [need to update uic pf version](https://github.com/ManageIQ/manageiq-ui-classic/pull/4253)

**Comment by AllenBW** [comment]
Cuz https://github.com/ManageIQ/manageiq-ui-classic/pull/4253 made it in, we're able to update the date time picker

@vconzola can yah check the updated screen shots above?

**Comment by vconzola** [comment]
@AllenBW New date/time picker looks good to me. Two minor things I see... (1) the Schedule button should be disabled until there's actually a date/time selected, and (2) after discussing further with other UXD folks I think we should standardize on scheduling being based on UTC time. So we should probably put some text to that affect in the UI. Let's talk about how to do this in our standup this afternoon.

**Comment by AllenBW** [comment]
https://github.com/ManageIQ/manageiq/pull/17672
https://github.com/ManageIQ/manageiq-api/pull/412

and another yet to be authored pr are the api ones we're waiting on... GOOD NEWS IS the end is in sight!

**Comment by priley86** [comment]
Will take a look at this again after lunch 🕐 ... looks like backend is ready?

**Comment by AllenBW** [comment]
Almost there @priley86 need to make sure delete works, working it presently, will add yah for review once its outta wip 👍

**Comment by AllenBW** [comment]
☝️ @priley86 @michaelkro yah want me to update it?

**Comment by michaelkro** [comment]
@AllenBW it looks like we do the same thing in our [bootstrap select component](https://github.com/ManageIQ/miq_v2v_ui_plugin/blob/master/app/javascript/react/screens/App/common/forms/BootstrapSelect.js#L7-L25).  (jquery setup in `componentDidMount`)

Seems like the way to go

**Comment by AllenBW** [comment]
😭 no functional component then, updated!

@michaelkro thanks for the thoroughness of review 🙇 ❤️ 🌮

**Comment by priley86** [comment]
Tested this locally - works fine for me. I did note a slight lag after scheduling (seems we are waiting for the next poll to update the list view?). Not a showstopper by any stretch...would like to get QE feedback.

**Comment by AllenBW** [comment]
@priley86 yeah its the polling, suppose we could force a fetch of not started migrations if yah would like though?

**Comment by priley86** [comment]
If it's not much additional work, fine w/ me ;) 

otherwise, lgtm

**Comment by AllenBW** [comment]
Yah, it'll be some work, can do it later if desired by QE 👍

**Comment by AparnaKarve** [comment]
@AllenBW The `"Validation failed: Name has already been taken"` issue described here https://github.com/ManageIQ/miq_v2v_ui_plugin/pull/401#issuecomment-399178902 still persists. Is that expected?

**Comment by AparnaKarve** [comment]
@AllenBW This looks really great... the delay however...

The delay is noticeable and big enough to get us in this situation -
An enabled Schedule button when the Plan is already scheduled.

screen shot 2018-07-11 at 4 39 03 pm


Clicking on the button causes the 500 error `Validation failed: Name has already been taken"`

Would it be possible to gray out the Schedule and the Unschedule button until the Plan is actually scheduled and unscheduled?

**Comment by AllenBW** [comment]
@AparnaKarve the notification is thrown as soon as the callback is resolved, as in when the plan is actually Scheduled/unscheduled. The delay is because we poll the page... [If the behavior blocks the merging of this pr, I can work on this nowish](https://github.com/ManageIQ/miq_v2v_ui_plugin/pull/401#issuecomment-404324430) 🤷‍♀️

**Comment by AllenBW** [comment]
So a few changes come to mind.... we can... (in order of effort, least to most to absolute least)

1. [increase the polling frequency here](https://github.com/ManageIQ/miq_v2v_ui_plugin/blob/eb1d28590dbc61119a9776263861c688a12a8f12/app/javascript/react/screens/App/Overview/Overview.js#L143) (this is only polling transformation plans, might make the whole migration process feel a little better)
2. swap the "schedule migration" button visibility to an internally managed variable rather than the presence or absence of the a schedule run_time (but this will still appear laggy, meaning the schedule time won't show up till polling happens)
3. pass `fetchTransformationPlansAction` all the way down into the modal, and throw that on click (but again, this might resolve before the schedule is completed, so not a guaranteed fix)
4. somehow call `fetchTransformationPlansAction` from the `scheduleMigration`(action) callback? feels really wrong to be calling an action from an action though... unsure if this is kosher 🤔 
5. give up and eat 🥑 🍿 by the 🌆 while wearing 🕴

**Comment by michaelkro** [comment]
> 4. somehow call fetchTransformationPlansAction from the scheduleMigration(action) callback? feels really wrong to be calling an action from an action though... unsure if this is kosher 🤔

I believe this is indeed kosher :)  One of the great things about `redux-thunk` is that it gives us total control over the `dispatch` process, and one of the things it allows us to do is dispatch multiples actions from a single action creator [(documentation)](https://github.com/reduxjs/redux-thunk#composition)

So, if you decided to go down this road, I will not complain :)

**Comment by priley86** [comment]
yes - of the options you list, it seems like 3 or 4 (calling `fetchTransformationPlansAction` again explicitly) should work. You can import actions (and reuse them) from within a nested view like [this](https://github.com/ManageIQ/miq_v2v_ui_plugin/blob/master/app/javascript/react/screens/App/Overview/screens/PlanWizard/components/PlanWizardGeneralStep/index.js#L3).

I should add that I don't think option 1 is not really an option - b/c I have heard previously that the backend automate code is only refreshing every 15 seconds, currently...so polling more frequently on the frontend seems like a bad route.

**Comment by AllenBW** [comment]
@michaelkro Thanks for the input! I've been chewing on what ya posted, still very unclear on how this approach might be successfully wired up. If you have some spare cycles would appreciate a pair on this :-/

**Comment by bdunne** [comment]
@AllenBW @AparnaKarve This should fix the "Name has already been taken" validation error.  https://github.com/ManageIQ/manageiq/pull/17696

**Comment by AllenBW** [comment]
@AparnaKarve, @michaelkro put up with me for 30 minutes and we have da solution, MORE DUPLICATE CODE EVERYWHERE aka, when you schedule or unschedule we refetch plans 😋 👍

**Comment by priley86** [comment]
@AllenBW sure thing... will test it out now.

**Comment by priley86** [comment]
yes - calling `fetchTransformationPlansAction` after `scheduleMigration` seems to have the desired effect... approving/merging to get this out for QE.

I am going to follow up w/ some schema validation checks later on... (always just a bonus).

**Comment by AparnaKarve** [comment]
@simaishi We will need this backported first before #491.
Thanks.

**Comment by simaishi** [comment]
and it looks like https://github.com/ManageIQ/manageiq-api/pull/400 needs to be backported before this PR...  btw, it will really help if all dependencies are listed in the top description, as I could easily miss if it's in comment somewhere.

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit 4032ca240376bd32d0673ce248459fc52e906f75
Author: Patrick Riley <priley86@gmail.com>
Date:   Thu Jul 12 10:44:34 2018 -0400

    Merge pull request #401 from AllenBW/enhancement/#369-schedule-migration
    
    [#369] Inactive migration plans can be scheduled to run
    (cherry picked from commit 18a306ff7a4767ca80d8a72b649e545a8afec683)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1608351
```

**Pull request merged**